editor/debugger/debug_adapter/debug_adapter_protocol.cpp

```diff
@@ -671,12 +674,150 @@ int DebugAdapterProtocol::parse_variant(const Variant &p_var) {
+	DAP::Variable node_type;
+	Array node_properties;
+
+	// Lambda to finish the current category
+	auto finishCategory = [&node_properties, &node_type, &properties, this]() {
+		if (!node_properties.is_empty()) {
+			node_type.value = itos(node_properties.size());
+			variable_list.insert(node_type.variablesReference, node_properties.duplicate());
+			properties.push_back(node_type.to_json());
+		}
+	};
+
+	for (SceneDebuggerObject::SceneDebuggerProperty &property : p_obj.properties) {
+		PropertyInfo &info = property.first;
+
+		// Script members ("Members/" prefix)
+		if (info.name.begins_with("Members/")) {
+			info.name = info.name.trim_prefix("Members/");
+			script_members.push_back(parse_object_variable(property));
+		}
+
+		// Script constants ("Constants/" prefix)
+		else if (info.name.begins_with("Constants/")) {
+			info.name = info.name.trim_prefix("Constants/");
+			script_constants.push_back(parse_object_variable(property));
+		}
+
+		// Script node ("Node/" prefix)
+		else if (info.name.begins_with("Node/")) {
+			info.name = info.name.trim_prefix("Node/");
+			script_node.push_back(parse_object_variable(property));
+		}
+
+		// Regular categories (with type Variant::NIL)
+		else if (info.type == Variant::NIL) {
+			finishCategory();
+
+			node_type.name = info.name;
+			node_type.type = "Category";
+			node_type.variablesReference = variable_id++;
+			node_properties.clear();
+		}
+
+		// Regular properties
+		else {
+			node_properties.push_back(parse_object_variable(property));
+		}
+	}
+
+	// Add the last category
+	finishCategory();
+
+	// Add the script categories, in reverse order to be at the front of the array
+	// ( [members; constants; node; category1; category2; ...] )
+	auto addScriptCategories = [&properties, this](const Array &p_properties, const String &p_name) {
+		if (!p_properties.is_empty()) {
+			DAP::Variable var;
+			var.name = p_name;
+			var.type = "Category";
+			var.value = itos(p_properties.size());
+			var.variablesReference = variable_id++;
+			variable_list.insert(var.variablesReference, p_properties);
+			properties.push_front(var.to_json());
+		}
+	};
+
+	addScriptCategories(script_node, "Node");
+	addScriptCategories(script_constants, "Constants");
+	addScriptCategories(script_members, "Members");
+
+	ERR_FAIL_COND(!object_list.has(object_id));
+	variable_list.insert(object_list[object_id], properties);
+}
+
+const Variant DebugAdapterProtocol::parse_object_variable(const SceneDebuggerObject::SceneDebuggerProperty &p_property) {
+	auto &[info, value] = p_property;
+
+	DAP::Variable var;
+	var.name = info.name;
+	var.type = Variant::get_type_name(info.type);
+	var.value = value;
+	var.variablesReference = parse_variant(value);
+
+	return var.to_json();
+}
+
+std::optional<ObjectID> DebugAdapterProtocol::search_object_id(int p_var_id) {
+	for (const auto &[object_id, var_id] : object_list) {
+		if (var_id == p_var_id) {
+			return object_id;
+		}
+	}
+	return std::nullopt;
+}
```

Comment:
`auto` is [not allowed](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html#auto-keyword)

Reply:
The documentation leads me to believe it was written only on the context of automatic type inference. For this feature, there is no option but to use `auto`, due to I assume how this was implemented in the standard.

Does that "zero `auto`" policy also apply in this context? I'm fine with it, but I think such a note on the docs would be important as well.

Reply:
Using the explicit key type is more clear, it makes it clear what types are available, that's the reason for the ban on `auto`, you can absolutely do this without `auto`, as per the suggestion

Reply:
I see, but I'd argue, in this context, providing information on the variable content is more important than it's type. The type can ultimately be checked on its definition, but the meaning of these variables is much harder to understand without explicit documentation.

As an example, if `ObjectID` didn't exist, then this would very likely be a `HashMap<int, int> object_list`; it would be confusing to know what these two `int`s refer to without a more thorough review (and especially here, since one refers to internal Godot IDs, and the other to DAP IDs; worst case, it could also introduce subtle bugs).

I still think it's worth it to be explicit here on the variable's content. What do you think of the following compromise?

```suggestion
ObjectID DebugAdapterProtocol::search_object_id(int p_var_id) {
	for (const KeyValue<ObjectID, int> &E: object_list) {
		const ObjectID& object_id = E.key;
		const int& var_id = E.value;
		
		if (var_id == p_var_id) {
			return object_id;
		}
	}
	return ObjectID();
}
```

Reply:
I don't see any reason to complicate things, and adding an indirection for three lines of code is a big excessive IMO, and it makes it more confusing IMO as it makes it harder to tell where this is coming from, please just keep it simple, no need to make it more complicated than it needs to be

Reply:
But ObjectID does exists, is you want to further clarify the `int` you can use something like:

```c++
// header file
using DAPVarID = int;
HashMap<ObjectID, DAPVarID> object_list;

// cpp file
for (const KeyValue<ObjectID, DAPVarID> &E: object_list) {
  // ....
}
```

Reply:
Huuum okay, I wasn't sure if `using ...` was allowed in the codebase; in that case, I agree with this approach.